File: queue/go/queue.go
package main

// Queue is an implementation of a queue
// using a doubly-linked list.
type Queue struct {
	linkedList doublyLinkedList
}

// Enqueue adds an element to the back of the queue.
func (q Queue) Enqueue(element interface{}) {
	q.linkedList.AddLast(element)
}

// Dequeue removes the element at the front of the queue
// and returns it. It returns an error if the queue is empty.
func (q Queue) Dequeue() (interface{}, error) {
	return q.linkedList.RemoveFirst()
}

// Peek returns the element at the front of the queue.
// It returns an error if the queue is empty.
func (q Queue) Peek() (interface{}, error) {
	return q.linkedList.Peek()
}

type doublyLinkedList interface {
	AddLast(element interface{})
	RemoveFirst() (interface{}, error)
	Peek() (interface{}, error)
}

// NewQueue returns a new Queue.
func NewQueue() Queue {
	return Queue{
		// linkedList: NewDoublyLinkedList(),
	}
}

File: queue/java/Queue.java
import java.util.NoSuchElementException;

/**
 * @author Ayomide Oyekanmi
 * @notes This implementation uses a Doubly Linked List as the underlying
 *          data structure
 */
class Queue<E> {

    // check linkedlist/doubly/java/DoublyLinkedList.java for implementation
    DoublyLinkedList<E> linkedList;

    public Queue(){
        linkedList = new DoublyLinkedList();
    }

    /**
     * Add element to back of the queue
     * 
     * @param element element to be added to the back of the queue
     */
    public void enqueue(E element){
        linkedList.addLast(element);
    }

    /**
     * Removes element at the front of the queue
     * 
     * @return returns removed element
     * @throws NoSuchElementException if queue is empty
     */
    public E dequeue()throws NoSuchElementException {
        return linkedList.removeFirst();
    }

    /**
     * Retrieves element at the front of the queue
     * 
     * @return returns retrieved element
     * @throws NoSuchElementException if queue is empty
     */
    public E peek()throws NoSuchElementException{
        return linkedList.peek();
    }
}